### PR TITLE
(todo completed) optimized posterize() by 1.65x and made it differentiable 

### PR DESCRIPTION
**Pull request by sidd-27**
The forward pass performs true posterization using uint8 and bit masking. The backward pass uses a straight-through estimator (STE), passing gradients as if the function were identity. This allows gradients to flow during training, enabling optimization despite non-differentiable operations.

https://colab.research.google.com/drive/10A13lay8GFLrTvl11j0Gzuhgsq3TzCAq?usp=sharing

here's a benchmark that shows speed improvement, as well as shows that gradients pass through the function

- [x] 🔬 New feature (non-breaking change which adds functionality)
- [x] optimization

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

**Comment by shijianjian** [comment]
I have a STE implemented here: https://github.com/kornia/kornia/blob/main/kornia/grad_estimator/ste.py, it is better if it can be reused

File: kornia/enhance/adjust.py [hunk 1]
```diff
@@ -785,6 +779,7 @@ def _mask(x: Tensor, shift: Tensor) -> Tensor:
             out = torch.zeros_like(input)
         elif b == 8:
             out = input.clone()
+            return STEFunction.apply(input.detach(), out, None)
```
Comment (edgarriba) [inline kornia/enhance/adjust.py:782]: i don't understand this

File: kornia/enhance/adjust.py [hunk 2]
```diff
@@ -732,23 +732,17 @@ def solarize(
 
 
 def posterize(input: Tensor, bits: Union[int, Tensor]) -> Tensor:
-    r"""Reduce the number of bits for each color channel.
-
-    .. image:: _static/img/posterize.png
```
Comment (edgarriba) [inline kornia/enhance/adjust.py:737]: why this is removed

**Review by edgarriba — CHANGES_REQUESTED** [review]

**Comment by edgarriba** [comment]
We are doing cleanup. Please, rebase and reopen if you still have interest in merging this PR.
